### PR TITLE
[BEAM-3713] Add pytest for unit tests

### PR DESCRIPTION
**Pull request by udim**
This is #7949 without IT support.

This does not replace regular precommit/postcommit/tox environments.
Pytest will replace nose once we can verify all tests are collected by pytest and blocking bugs are solved (see BEAM-3713).

- Runs unit tests using pytest
  - tox: `tox -e py27-gcp-pytest,py36-pytest,etc.`
  - gradle: `../../gradlew pythonPreCommitPytest`
  - github PR phrase: `run python_pytest precommit`
- Tests run in parallel (still single-threaded but on multiple worker
processes).
  - no_xdist marker used for tests that don't work the xdist plugin.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by udim** [comment]
run seed job

**Comment by udim** [comment]
run seed job

**Comment by udim** [comment]
R: @chadrik

This should be tested with a "run seed job", which should succeed. "run python_pytest precommit" will then launch a pytest-based precommit job (without ITs), but it's not critical that this job passes at this stage.

**Comment by chadrik** [comment]
This looks great.  What's the plan for deploying it?  Do you plan to merge it as is to get it in front of more users for testing, or will you replace nose with pytest before the merge?

**Comment by chadrik** [comment]
Sounds good.

File: sdks/python/apache_beam/runners/dataflow/dataflow_runner_test.py [hunk 1]
```diff
@@ -228,6 +229,8 @@ def test_biqquery_read_streaming_fail(self):
                                  r'source is not currently available'):
       p.run()
 
+  @pytest.mark.skipif(sys.version_info >= (3, 7),
+                      reason='TODO(BEAM-8095): Segfaults in Python 3.7')
```
Comment (chadrik) [inline sdks/python/apache_beam/runners/dataflow/dataflow_runner_test.py]: hmmm... this only segfaults when using pytest?   what happens if you disable xdist?

Comment (udim) [reply sdks/python/apache_beam/runners/dataflow/dataflow_runner_test.py]: I changed it to `no_xdist`. It still segfaults when using xdist or when attempting to run under a debugger in PyCharm.

**Comment by udim** [comment]
@chadrik PTAL
CC/R: @markflyhigh, @chamikaramj

File: sdks/python/conftest.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+import sys
+
+# See pytest.ini for main collection rules.
+collect_ignore_glob = []
+if sys.version_info < (3,):
+  collect_ignore_glob.append('*_py3.py')
+for minor in [5, 6, 7, 8, 9]:
+  if sys.version_info < (3, minor):
+    collect_ignore_glob.append('*_py3%d.py' % minor)
```
Comment (chadrik) [inline sdks/python/conftest.py]: How about something like this:

```python
MAX_SUPPORTED_PYTHON_VERSION = (3, 8)

# See pytest.ini for main collection rules.
collect_ignore_glob = []
if sys.version_info < (3,):
  collect_ignore_glob.append('*_py3.py')
else:
  for minor in range(sys.version_info.minor +1, MAX_SUPPORTED_PYTHON_VERSION[1] + 1):
    collect_ignore_glob.append('*_py3%d.py' % minor)
```

**Comment by chadrik** [comment]
I had two thoughts related to this today, as I struggled trying to diagnose some failed tests on jenkins.

1. it would be great if you could make the gradle tasks run serially, since we're already using xdist to parallelize the work.  That will make the logs easier to read since it won't interleave different tox tasks together.
2. I added a PR to enable color in tox: #9840.  Hopefully this will work with pytest, because I think it will add a lot of value (unbelievably, nose does not support color out of the box).

**Comment by udim** [comment]
> I had two thoughts related to this today, as I struggled trying to diagnose some failed tests on jenkins.
> 
> 1. it would be great if you could make the gradle tasks run serially, since we're already using xdist to parallelize the work.  That will make the logs easier to read since it won't interleave different tox tasks together.
Yes that makes sense, the logs already make my browser slow down to a crawl (#9813). Let's do that in another PR though.

> 2. I added a PR to enable color in tox: #9840.  Hopefully this will work with pytest, because I think it will add a lot of value (unbelievably, nose does not support color out of the box).

Colors already work for me with tox and pytest. Let's get this PR merged so you can try it out yourself. :)

**Comment by chadrik** [comment]
> Colors already work for me with tox and pytest. 
Great.  I've been digging into my problem with colored mypy output problem and I think it's a tty issue.  I think mypy needs a mechanism to force colors even when a tty is not present.

>> it would be great if you could make the gradle tasks run serially, since we're already using xdist to parallelize the work. That will make the logs easier to read since it won't interleave different tox tasks together.

> Yes that makes sense, the logs already make my browser slow down to a crawl (#9813). Let's do that in another PR though.

I've also been thinking about investigating using tox's support fo parallel execution instead of gradle: https://tox.readthedocs.io/en/latest/example/basic.html#parallel-mode

> Let's get this PR merged so you can try it out yourself. :)

Looks good to me!

File: sdks/python/conftest.py (hunk 2 above)
Comment (chadrik) [reply sdks/python/conftest.py]: oh, btw, what did you think of this suggestion?

Comment (udim) [reply sdks/python/conftest.py]: IDK. It's more verbose but I don't have a preference.

Ideally, I'd like a single source of truth for the list of Python supported versions (`python_requires` in setup.py is such a source but it's hard to parse).

**Comment by udim** [comment]
PTAL

File: sdks/python/conftest.py (hunk 2 above)
Comment (chadrik) [reply sdks/python/conftest.py]: > IDK. It's more verbose but I don't have a preference.

I think it'll be easier to maintain long term: just update the max support python version.  

But maintainability aside,  it seems a bit odd to test for our lower bound in a loop when we know precisely what the lower bound is: `sys.version_info.minor +1`

> Ideally, I'd like a single source of truth for the list of Python supported versions (python_requires in setup.py is such a source but it's hard to parse).

true.

Comment (udim) [reply sdks/python/conftest.py]: > But maintainability aside, it seems a bit odd to test for our lower bound in a loop when we know precisely what the lower bound is: `sys.version_info.minor +1`

You're right.

File: sdks/python/apache_beam/runners/dataflow/dataflow_runner_test.py (hunk 1 above)
Comment (tvalentyn) [reply sdks/python/apache_beam/runners/dataflow/dataflow_runner_test.py]: For problems with this test see: https://issues.apache.org/jira/browse/BEAM-8397?focusedCommentId=16958087&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-16958087

**Comment by chadrik** [comment]
Looks good.  I say we merge this once the tests are confirmed passing.

**Comment by udim** [comment]
Run Java PreCommit

**Comment by udim** [comment]
Run CommunityMetrics PreCommit

**Comment by udim** [comment]
Run Python PreCommit

**Comment by udim** [comment]
Run Python PreCommit

**Comment by udim** [comment]
Merging since tests pass. Will address BEAM-8095 TODO in a subsequent PR. I believe it has been fixed in #9881.
